File: src/utils/holiday.ts
/**
 * 节假日服务模块
 * 通过 Tauri 后端调用 NateScarlet/holiday-cn 开源数据获取中国法定节假日
 */
import { invoke } from '@tauri-apps/api/core'

export interface HolidayInfo {
  /** 日期 YYYY-MM-DD */
  date: string
  /** 是否是休息日 */
  isHoliday: boolean
  /** 是否是工作日（包括调休） */
  isWorkday: boolean
  /** 是否是调休日（周末需要上班） */
  isAdjust: boolean
  /** 节日名称（如：春节、国庆节），非节日则为空 */
  name: string
  /** 节日类型：1-法定节假日 2-调休上班 */
  type: number
}

// 后端返回的节假日数据格式
interface BackendHolidayInfo {
  date: string
  name: string
  isOffDay: boolean
}

// 年度节假日缓存
const holidayCache = new Map<number, Map<string, HolidayInfo>>()

/**
 * 从 Tauri 后端获取指定年份的节假日数据
 * @param year 年份
 */
async function fetchYearHolidays(year: number): Promise<Map<string, HolidayInfo>> {
  try {
    const holidays = await invoke<BackendHolidayInfo[]>('fetch_holidays', { year })
    const result = new Map<string, HolidayInfo>()
    
    for (const info of holidays) {
      result.set(info.date, {
        date: info.date,
        isHoliday: info.isOffDay,
        isWorkday: !info.isOffDay,
        isAdjust: !info.isOffDay, // 如果不是休息日但在节假日数据中，说明是调休上班
        name: info.name,
        type: info.isOffDay ? 1 : 2
      })
    }
    
    return result
  } catch (error) {
    console.error(`Failed to fetch holidays for ${year}:`, error)
    return new Map()
  }
}

/**
 * 获取指定年份的节假日数据（带缓存）
 * @param year 年份
 */
export async function getYearHolidays(year: number): Promise<Map<string, HolidayInfo>> {
  // 检查缓存
  if (holidayCache.has(year)) {
    return holidayCache.get(year)!
  }
  
  // 从 API 获取
  const holidays = await fetchYearHolidays(year)
  
  // 存入缓存
  if (holidays.size > 0) {
    holidayCache.set(year, holidays)
  }
  
  return holidays
}

File: src-tauri/src/commands/holiday.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// 节假日信息
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct HolidayInfo {
    pub date: String,
    pub name: String,
    pub is_off_day: bool,
}

/// NateScarlet/holiday-cn 数据格式
/// 数据源: https://raw.githubusercontent.com/NateScarlet/holiday-cn/master/{year}.json
#[derive(Debug, Deserialize)]
struct HolidayCnResponse {
    #[serde(rename = "$schema")]
    #[allow(dead_code)]
    schema: Option<String>,
    #[serde(rename = "$id")]
    #[allow(dead_code)]
    id: Option<String>,
    #[allow(dead_code)]
    year: i32,
    #[allow(dead_code)]
    papers: Option<Vec<serde_json::Value>>,
    days: Vec<HolidayCnDay>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct HolidayCnDay {
    name: String,
    date: String,
    is_off_day: bool,
}

/// 获取节假日缓存目录
fn get_cache_dir() -> PathBuf {
    dirs::data_local_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("mini-todo")
        .join("holidays")
}

/// 获取指定年份的缓存文件路径
fn get_cache_path(year: i32) -> PathBuf {
    get_cache_dir().join(format!("{}.json", year))
}

/// 从本地缓存读取节假日数据
fn read_cache(year: i32) -> Option<Vec<HolidayInfo>> {
    let path = get_cache_path(year);
    let data = std::fs::read_to_string(&path).ok()?;
    serde_json::from_str(&data).ok()
}

/// 将节假日数据写入本地缓存
fn write_cache(year: i32, holidays: &[HolidayInfo]) {
    let cache_dir = get_cache_dir();
    let _ = std::fs::create_dir_all(&cache_dir);
    if let Ok(data) = serde_json::to_string(holidays) {
        let _ = std::fs::write(get_cache_path(year), data);
    }
}

/// 从网络获取节假日数据（带超时）
async fn fetch_from_network(year: i32) -> Result<Vec<HolidayInfo>, String> {
    let url = format!(
        "https://raw.githubusercontent.com/NateScarlet/holiday-cn/master/{}.json",
        year
    );

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .connect_timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {}", e))?;

    let response = client
        .get(&url)
        .send()
        .await
        .map_err(|e| format!("Failed to fetch holidays: {}", e))?;

    if !response.status().is_success() {
        return Err(format!("API returned status: {}", response.status()));
    }

    let api_response: HolidayCnResponse = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse response: {}", e))?;

    let holidays: Vec<HolidayInfo> = api_response
        .days
        .into_iter()
        .map(|day| HolidayInfo {
            date: day.date,
            name: day.name,
            is_off_day: day.is_off_day,
        })
        .collect();

    Ok(holidays)
}

/// 获取指定年份的节假日数据
/// 策略：优先从网络获取（带超时），失败时回退到本地缓存
/// 网络获取成功后自动更新本地缓存
#[tauri::command]
pub async fn fetch_holidays(year: i32) -> Result<Vec<HolidayInfo>, String> {
    // 尝试从网络获取
    match fetch_from_network(year).await {
        Ok(holidays) => {
            // 网络获取成功，更新本地缓存
            write_cache(year, &holidays);
            Ok(holidays)
        }
        Err(network_err) => {
            // 网络获取失败，尝试从本地缓存读取
            eprintln!("Network fetch failed for year {}: {}", year, network_err);
            if let Some(cached) = read_cache(year) {
                eprintln!("Using cached holiday data for year {}", year);
                Ok(cached)
            } else {
                Err(format!(
                    "Failed to fetch holidays and no local cache: {}",
                    network_err
                ))
            }
        }
    }
}
